docs: add Node app guide (#13995)
web-infra-dev/rspack: website/docs/en/guide/tech/nestjs.mdx

---
description: 'Use NestJS with Rspack, including setup, TypeScript decorators, development HMR, and production minification.'
---

import { PackageManagerTabs } from '@theme';

# NestJS

> Rspack can build [NestJS](https://nestjs.com/) applications with the same Node.js runtime setup used by other server applications. Compared with `tsc`, Rspack can bundle application code, transform TypeScript with SWC, enable development HMR, and leave runtime dependencies external when needed.

Reviewer: ```suggestion
Rspack can build [NestJS](https://nestjs.com/) applications. Compared with `tsc`, Rspack can bundle application code, transform TypeScript with SWC, enable development HMR, and leave runtime dependencies external when needed.
```

## How to use

Start with the common [Node.js application configuration](/guide/tech/node), then add the NestJS settings described below. You can also use the [NestJS example](https://github.com/rstackjs/rstack-examples/tree/main/rspack/nestjs) in `rstack-examples` as a reference. It includes a complete `rspack.config.mjs`, development script, production build script, and HMR entry.

## Example

The [rstack-examples](https://github.com/rstackjs/rstack-examples/tree/main/rspack/nestjs) repository contains a complete NestJS application built with Rspack.

## Install dependencies

Install the common Node.js build dependencies from the [Node.js guide](/guide/tech/node#install-dependencies), then install the NestJS runtime packages:

<PackageManagerTabs command="add @nestjs/common @nestjs/core @nestjs/platform-express reflect-metadata rxjs" />

## Basic concepts

> NestJS applications run in Node.js, so the runtime target, external dependencies, development scripts, and native addon handling can follow the common [Node.js configuration](/guide/tech/node). The main NestJS-specific requirement is decorator metadata, which NestJS uses at runtime.

Reviewer: ```suggestion
NestJS applications run in Node.js, so most configurations can follow the common [Node.js configuration](/guide/tech/node). The main NestJS-specific requirement is decorator metadata, which NestJS uses at runtime.
```
